woocommerce/woocommerce
# Validate and present custom error for not in allowed emails coupons

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR introduces upfront validation in allowed emails coupons, opposed to the current procedure of adding the coupons and removing them when placing the Order if the billing email does not match the coupon whitelist of emails.
It also updates the error messages presented to the user, stating clearly why the coupon was not accepted and how to proceed to use it effectively.

| |Cart | Checkout |
|-|-|-|
| Blocks | <img width="250" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/46eaac48-e979-4f8b-917f-6b0a8558a729"> | <img width="250" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/dffa7196-efbb-43fd-b68e-cacb32e04637"> |
| Shortcode | <img width="250" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/23976dac-a6a1-4a01-b130-5d11c5e05b11"> |<img width="250" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/ae055cd3-39ac-49c1-a035-01680278ee4e"> |


Closes #26289
Closes #26305



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prior to testing
1. Create four pages with Shortcode Cart, Shortcode Checkout, Blocks Cart and Blocks Checkout
2. In WC Settings, set the Cart page to the Shortcode Cart page you created (we need this to avoid the bug reported here https://github.com/woocommerce/woocommerce/issues/44232 )
3. Go to Marketing > Coupons
4. Create a new coupon restricted to the email allow@yay.com (make sure this email is not associated with your account), with any discount amount
5. Create another coupon named "test" without restrictions, with any discount amount
6. Add some items to your cart.

#### Shortcode Cart Test
1. Go to the shortcode cart page
2. Apply the coupon "test" and verify it gets applied and the totals update
3. Apply the restricted coupon and verify it doesn't get applied, and the proper error notice is displayed

#### Blocks Cart test
1. Go to the Blocks Cart page
2. Apply the coupon "test" and verify it gets applied and the totals update
3. Apply the restricted coupon and verify it doesn't get applied, and the proper error is displayed

#### Shortcode Checkout test
1. Go to the shortcode Checkout page
2. Fill the email field to something else than "allow@yay.com"
3. Apply the restricted coupon and verify it doesn't get applied, and the proper error notice is displayed
4. Change the email field to "allow@yay.com"
5. Apply the restricted coupon and verify it gets applied correctly
6. Place the order and verify applied discounts were effective

#### Blocks Checkout test
1. Go to the Blocks Checkout page
2. Fill the email field to something else than "allow@yay.com"
3. Apply the restricted coupon and verify it doesn't get applied, and the proper error is displayed
4. Change the email field to "allow@yay.com"
5. Apply the restricted coupon and verify it gets applied correctly
6. Place the order and verify applied discounts were effective


### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
